microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceDiagnosticsParticipant.java

```diff
@@ -0,0 +1,151 @@
+import static org.eclipse.lsp4mp.jdt.internal.faulttolerance.MicroProfileFaultToleranceConstants.FALLBACK_ANNOTATION_ABBR;
+import static org.eclipse.lsp4mp.jdt.internal.faulttolerance.java.MicroProfileFaultToleranceErrorCode.FALLBACK_METHOD_DOES_NOT_EXIST;
+import static org.eclipse.lsp4mp.jdt.internal.faulttolerance.MicroProfileFaultToleranceConstants.FALLBACK_METHOD_FALLBACK_ANNOTATION_MEMBER;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.jdt.core.IJavaElement;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.IOpenable;
+import org.eclipse.jdt.core.IType;
+import org.eclipse.jdt.core.ITypeRoot;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.core.dom.ASTVisitor;
+import org.eclipse.jdt.core.dom.CompilationUnit;
+import org.eclipse.jdt.core.dom.Expression;
+import org.eclipse.jdt.core.dom.MemberValuePair;
+import org.eclipse.jdt.core.dom.MethodDeclaration;
+import org.eclipse.jdt.core.dom.NormalAnnotation;
+import org.eclipse.jdt.internal.core.manipulation.dom.ASTResolving;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.DiagnosticSeverity;
+import org.eclipse.lsp4mp.jdt.core.java.diagnostics.IJavaDiagnosticsParticipant;
+import org.eclipse.lsp4mp.jdt.core.java.diagnostics.JavaDiagnosticsContext;
+import org.eclipse.lsp4mp.jdt.core.utils.JDTTypeUtils;
+
+public class MicroProfileFaultToleranceDiagnosticsParticipant implements IJavaDiagnosticsParticipant {
+	
+	@Override
+	public boolean isAdaptedForDiagnostics(JavaDiagnosticsContext context, IProgressMonitor monitor)
+			throws CoreException {
+		IJavaProject javaProject = context.getJavaProject();
+		return JDTTypeUtils.findType(javaProject, FALLBACK_ANNOTATION) != null;
+	}
+
+	@Override
+	public List<Diagnostic> collectDiagnostics(JavaDiagnosticsContext context, IProgressMonitor monitor)
+			throws CoreException {
+		ITypeRoot root = context.getTypeRoot();
+		IJavaElement[] elements = root.getChildren();
+		List<Diagnostic> diagnostics = new ArrayList<>();
+		collectDiagnostics(elements, diagnostics, context, monitor);
+		return diagnostics;
+	}
+
+	private void collectDiagnostics(IJavaElement[] elements, List<Diagnostic> diagnostics,
+			JavaDiagnosticsContext context, IProgressMonitor monitor) {
+		for (IJavaElement element : elements) {
+			if (monitor.isCanceled()) {
+				return;
+			} else if (element.getElementType() == IJavaElement.TYPE) {
+				IType type = (IType) element;
+				validateClass(type, diagnostics, context, monitor);
+			}
+		}
+	}
+
+	private void validateClass(IType type, List<Diagnostic> diagnostics, JavaDiagnosticsContext context,
+			IProgressMonitor monitor) {
+		Set<String> methodSet;
```

Comment:
Your methodSet compute should be done in a static method and this method must be called only when there is a Fallback annotation in the class.

To do that I suggest you that you create an inner class like 

```java
static class MicroProfileFaultTolerantValidator extends ASTVisitor {
    private final IType type;
    private Set<String> existingMethods ;

    public MicroProfileFaultTolerantValidator( IType type) {
        this.type = type;
    }

    public boolean visit(MethodDeclaration node) {
	validateMethod(node, type, diagnostics, context, monitor);  
    }

    private void validateMethod(MethodDeclaration node, Set<String> methodSet, IType type, List<Diagnostic> diagnostics,
			JavaDiagnosticsContext context, IProgressMonitor monitor) {
      ....
      if (existingMethods == null) {
            existingMethods  = getExistingMethods(type);
      }
}
```